### ISSUE TITLE
Inking tool: add global simplification button

### ISSUE THREAD

**Issue by 0ion9** [mypaint/mypaint#224]
As I look at the waypoints generated by the Inking tool, and attempt to adjust them, one thing occurs to me:

There are too many of them to make adjustment easy.
This is a problem similar to the problem Inkscape's pencil tool has to deal with -- representing the path with sufficient nodes but not too many. So there may be something to be learnt from that.

In any case, what I'm suggesting is a simple post-processing heuristic, where, given points A B C, the angle AB is compared with the angle BC, and if it's similar enough, B is discarded. The simplification option would control the threshold used to determine 'similar enough'.
To avoid distortion, the amount of error created by this would need to be tracked(by cumulatively summing the AB vs BC difference during post processing), and if B would otherwise be discarded but error level is above a threshold, B is kept and the error value is reset to 0.

The intended result is like a type of adaptive tesselation -- more points are placed near changes of direction.

**Comment by achadwick** [comment]
(This refers to the newly added gui/inktool.py at eb76974)

Hmm, this sounds interesting. When exactly do you suggest it should happen and how many times? I guess it could be applied during the adjust phase proper by clicking a button...

Changing the spline control points will cause a re-rendering of the curve with the brush engine, and that can be slow and flickery.
I really want to avoid any flicker between ending the stroke that creates the initial curve and the buttons appearing. 
I really, really want to avoid any flicker during capture itself.

The ink tool currently just uses some rather ad-hoc time and distance metrics during the capture phase, but what you're describing sounds like it could drop nodes anywhere in the body of the curve at any time during capture.

**Comment by 0ion9** [comment]
It would be a postprocessing effect that would fire immediately after the path was completed, hopefully similar in effect to Inkscape's Ctrl+L Simplify. You could alter capture parameters to achieve similar effect, but I have less understanding of how to get a good result in that case.

Preferably, it would run before the rendering system renders the completed path. In reality, I guess that may not be possible due to the way it's rendered in an idle handler.

The algorithm I describe would only need one pass. 

Supposing your complete set of points are
A B C D E F G, and B and E are ultimately going to be removed (because their angle change is small). The algorithm would:

* examine A B C, compare AB and BC, remove B
* examine A C D, compare AC and CD
* examine C D E, compare CD and DE
* examine D E F, compare DE and EF, remove E
* examine E F G, compare EF and FG.

(Where two-letter symbols represent the angle between two points)

So, if I understand correctly what you are saying , this would cause some "flicker between ending the stroke that creates the initial curve and the buttons appearing." but no "flicker during capture itself."

It's possible that it could have some use as a separately invokable function. This does introduce additional complications, but would make flicker pretty excusable.

**Comment by jy2wong** [comment]
*puts on math major hat :tophat: * 
Exercise caution with the above algorithm.
Suppose we have the following arrangement:
```
A--------------------------------------------B-------C
                                                     |
                                                     |
                                                     D---------E
```
Every segment is determined by its end points, the point before the first endpoint and the point after the last endpoint. It's possible that by removing B, the segment CD behaves unexpectedly (or that the new segment AC behaves unexpectedly).

Also, just to be explicit, the code for the above algorithm would be
```python
i=0
while i<points.size()-2:
    while abs(angle(points[i], points[i+1]) - angle(points[i+1], points[i+2])) < THRESHOLD:
        points.pop(i+1)
    i++
```

**Comment by achadwick** [comment]
Just added an options widget for the ink tool allowing the raw input values of a selected node to be edited: commit a262359d20d90b0c9d0e142ba892da480c1f715e. It's a short step from there to a "delete node" button or global stuff like simplifying the whole of the current curve.

**Comment by achadwick** [comment]
I've added a button for removing the currently selected node. That should suffice for the 1.2 release; retitling and bumping.

**Comment by 0ion9** [comment]
Thanks. Now that I have access to deletion, I see that the continuity of curves is less than I thought, so deleting nodes often makes a path overly polygonal. I've posted a new issue about manipulating more than one point at once.

**Comment by achadwick** [comment]
> EDIT: BTW, the inking tool does not appear on the toolbar. Maybe it should, before 1.2?

Good job I'm still subscribed to this. You're looking for this icon:

![image](https://cloud.githubusercontent.com/assets/61299/8002369/151e895e-0b66-11e5-9be2-04cc224fefb3.png)

and it's considered a line mode in the toolbar's right button menu.

If it isn't there, please raise a separate issue about it.

**Comment by 0ion9** [comment]
Ah, NVM then, it's there

**Comment by Cloudef** [comment]
Just FYI, there exists simplification algorithms just for this purpose:
https://en.wikipedia.org/wiki/Ramer%E2%80%93Douglas%E2%80%93Peucker_algorithm

That is one I've used in past.

**Comment by achadwick** [comment]
@jy2wong , @Cloudef great information, thanks. Armed with that, I should be able to add something after 1.2.0 is out!

(Clean, tested patches or PRs/branches adding this would be really welcome then too :grin: )

**Comment by achadwick** [comment]
Closed with 1a9afb998b5c039114b1000196e6e3a4afef4a8d thru ce6e32ebd15ccf26e003b91d92f85898f1ee1996 extracted from #522. Thanks, @dothiko!

**Comment by charbelnicolas** [comment]
Hello, I'm switching from Krita to MyPaint and I noticed that the inking tool flickers too much. Is there a way to get rid of the flicker? Maybe show a placeholder curve while moving the edit points?
